std-bits/base/src/main/java/org/enso/base/Encoding_Utils.java

```diff
@@ -167,15 +168,13 @@ private static ReportingStreamDecoder create_stream_decoder(InputStream stream,
    * <p>It returns the result returned from the executed action and any encoding problems that
    * occurred when processing it.
    */
-  public static <R> WithProblems<R, String> with_stream_decoder(
-      InputStream stream, Charset charset, Function<ReportingStreamDecoder, R> action)
+  public static WithProblems<Value, String> with_stream_decoder(
+      InputStream stream, Charset charset, Function<ReportingStreamDecoder, Value> action)
       throws IOException {
-    R result;
+    Value result;
     ReportingStreamDecoder decoder = create_stream_decoder(stream, charset);
-    try {
+    try (decoder) {
```

Comment:
Has this the same meaning as
```java
try (ReportingStreamDecoder decoder = create_stream_decoder(stream, charset)) { ... }
```
that would be my preferred code style.

Reply:
Will change.

Reply:
Nvm, I cannot. Note that I'm using the `decoder` _after_ the `try` and this is on purpose.

I need to wait until _after_ I `close` the decoder before I read what problems it reported (since some problems may be reported during closing).

So I think this code must stay as-is.

Reply:
Using `decoder` after it has been `close()`! Isn't that logically unsound ;-?

An option might be to:
```java
try (ReportingStreamDecoder decoder = create_stream_decoder(stream, charset)) {
   return new WithProblems<>(result, decoder.getReportedProblems());
}
```
which would work the same unless `close()` influences the return value of `getReportedProblems()`.

Reply:
I wrote the `ReportingStreamDecoder` and designed it in this way.

As I said above, _yes_ `close` does influence the return value:

> (since some problems may be reported during closing).

So I _have_ to use it after closing, its part of its design to allow reporting warnings that may happen during that phase.

Reply:
I guess we could rewrite it to have some separate `WarningReported` reference, which can be used externally to report warnings the warnings that happen during closing. But I'm not sure if it's worth doing such a refactor right now.